lib/src/renamer.dart

```diff
@@ -0,0 +1,259 @@
+/// succeeds and may contain references to captured groups in the matcher's
+/// regular expression.
+///
+/// For an input, statements are evaluated in order until a match is found. Once
+/// a match is found, its output is returned immediately, bypassing any
+/// subsequent statements.
+///
+/// These clauses are combined as `<key> <matcher> to <output>`. If the key in
+/// question is the empty string, you omit that clause, so you get
+/// `<matcher> to <output>`.
+///
+/// If you wish to include a semicolon or space in the matcher or output clause,
+/// you can either escape it with `\` or wrap the entire clause in single or
+/// double quotes.
+class Renamer<T> {
+  /// A map from keys to functions that take an input and return the value of
+  /// that key for that input.
+  final Map<String, String Function(T input)> keys;
+
+  /// The list of statements that are evaluated in order by this renamer.
+  final List<_Statement<T>> _statements;
+
+  Renamer._(this.keys, this._statements);
+
+  /// Creates a simple Renamer from [code] that only uses an empty key.
+  static Renamer<String> simple(String code) {
+    return Renamer(code, {'': (input) => input});
+  }
+
+  /// Creates a Renamer from [code] that takes a map from keys to string values
+  /// as input.
+  ///
+  /// [keys] is the list of keys that can be referenced in [code] and must
+  /// all appear in every input map.
+  static Renamer<Map<String, String>> map(String code, List<String> keys) {
+    return Renamer(code, {for (var key in keys) key: (input) => input[key]});
+  }
+
+  /// Creates a Renamer based on [code] and a map from keys to functions that
+  /// take an input and return a string.
+  ///
+  /// All keys must consist of only lowercase letters, underscores, and hyphens.
+  ///
+  /// If provided, [sourceUrl] will appear in parsing errors. It can be
+  /// a [String] or a [Uri].
+  factory Renamer(String code, Map<String, String Function(T input)> keys,
+      {dynamic sourceUrl}) {
+    for (var key in keys.keys) {
+      if (!RegExp(r'^[a-z_-]*$').hasMatch(key)) {
+        throw ArgumentError(
+            'Invalid key "$key". Must use only lowercase letters, '
+            'underscores, and hyphens.');
+      }
+    }
+    var scanner = StringScanner(code, sourceUrl: sourceUrl);
+    var statements = <_Statement<T>>[];
+    scanner.scan(_statementDelimiter);
+    while (!scanner.isDone) {
+      statements.add(_readStatement(scanner, keys));
+    }
+    return Renamer._(keys, statements);
+  }
+
+  /// Reads the next statement (and the trailing delimiter, if any) from
+  /// [scanner].
+  static _Statement<T> _readStatement<T>(
+      StringScanner scanner, Map<String, String Function(T input)> keys) {
+    var start = scanner.position;
+    FormatException lastException;
+    // Tries each key in succession until one is successfully returned.
+    for (var entry in keys.entries) {
+      try {
```

Comment:
If I'm understanding correctly, the reason you have a `try`/`catch` here is to handle the case where a user uses the default key is used for a pattern that's also the same as a key name (e.g. `url to my-url`), right? That's a complex enough case that it's worth documenting in a comment (and testing if you aren't already).

For reference, catching exceptions like this to handle control flow is quite slow. This code is not performance-critical at all so it probably doesn't matter, but if you wanted to make something like this more efficient you could potentially do so by doing a "lexing" pass where you handle whitespace separators and escape sequences and return a list of strings in the next statement, and then a "parsing" phase where you can make decisions based on how many strings there are.

Reply:
Added a comment and a couple of tests.

I considered that kind of two-pass parse, but the matcher and output clauses have slightly different behaviors for escape sequences (since the matcher clause needs to leave most of them intact for the regex parser to handle, but the output clause handles them all). Because of that, I figured the performance cost of the exceptions was worth the simpler implementation given that only the parsing of the renamer scripts would be slowed, not the actual execution.